File: standalone/ws.go
package main

import (
	"database/sql"
	"encoding/json"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// Near-real-time sync between devices with the same pad open — port of
// server/ws.js. One WebSocket per tab, grouped by pad ("room"); when the
// pad changes, every connected client gets notified and re-fetches the
// current state from the server.

type wsHub struct {
	mu    sync.Mutex
	rooms map[string]map[*websocket.Conn]bool
}

func newWSHub() *wsHub {
	return &wsHub{rooms: map[string]map[*websocket.Conn]bool{}}
}

func (h *wsHub) join(padID string, c *websocket.Conn) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if h.rooms[padID] == nil {
		h.rooms[padID] = map[*websocket.Conn]bool{}
	}
	h.rooms[padID][c] = true
}

func (h *wsHub) leave(padID string, c *websocket.Conn) {
	h.mu.Lock()
	defer h.mu.Unlock()
	if room, ok := h.rooms[padID]; ok {
		delete(room, c)
		if len(room) == 0 {
			delete(h.rooms, padID)
		}
	}
}

func (h *wsHub) broadcastPadChanged(padID string, extra map[string]any) {
	h.mu.Lock()
	room := h.rooms[padID]
	conns := make([]*websocket.Conn, 0, len(room))
	for c := range room {
		conns = append(conns, c)
	}
	h.mu.Unlock()
	if len(conns) == 0 {
		return
	}
	payload := map[string]any{"type": "changed"}
	for k, v := range extra {
		payload[k] = v
	}
	b, _ := json.Marshal(payload)
	for _, c := range conns {
		_ = c.WriteMessage(websocket.TextMessage, b)
	}
}

var wsUpgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
	// No Origin check: the real authorization happens via signed cookie in
	// isWSRequestAuthorized, just like in the Node server.
	CheckOrigin: func(r *http.Request) bool { return true },
}

// isWSRequestAuthorized mirrors server/ws.js#isRequestAuthorized: reads the
// signed fp_site/fp_unlocked cookies directly from the upgrade request
// (the same mechanism the regular HTTP routes use).
func isWSRequestAuthorized(cfg *Config, db *sql.DB, r *http.Request, padID string) bool {
	if cfg.SitePassword != "" {
		c, err := r.Cookie(siteCookie)
		if err != nil {
			return false
		}
		val, ok := unsignValue(cfg.CookieSecret, c.Value)
		if !ok || val != "ok" {
			return false
		}
	}
	pad, err := getPad(db, padID)
	if err != nil {
		return false
	}
	if pad != nil && pad.PasswordHash.Valid && pad.PasswordHash.String != "" {
		if !getUnlockedSet(cfg, r)[padHash(padID)] {
			return false
		}
	}
	return true
}

func handleWS(cfg *Config, db *sql.DB, hub *wsHub) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		padID := normalizePadID(r.URL.Query().Get("pad"))
		if padID == "" {
			http.Error(w, "invalid pad", http.StatusBadRequest)
			return
		}
		if !isWSRequestAuthorized(cfg, db, r, padID) {
			http.Error(w, "unauthorized", http.StatusUnauthorized)
			return
		}
		conn, err := wsUpgrader.Upgrade(w, r, nil)
		if err != nil {
			return
		}
		hub.join(padID, conn)

		conn.SetReadDeadline(time.Now().Add(60 * time.Second))
		conn.SetPongHandler(func(string) error {
			conn.SetReadDeadline(time.Now().Add(60 * time.Second))
			return nil
		})

		done := make(chan struct{})
		go func() {
			defer close(done)
			for {
				if _, _, err := conn.ReadMessage(); err != nil {
					return
				}
			}
		}()

		ticker := time.NewTicker(30 * time.Second)
		defer ticker.Stop()
		for {
			select {
			case <-done:
				hub.leave(padID, conn)
				conn.Close()
				return
			case <-ticker.C:
				if err := conn.WriteControl(websocket.PingMessage, nil, time.Now().Add(5*time.Second)); err != nil {
					hub.leave(padID, conn)
					conn.Close()
					return
				}
			}
		}
	}
}

File: server/ws.js
'use strict';

/*
 * Near-real-time sync between devices that have the same pad open. One
 * WebSocket per tab, grouped by pad ("room"). When a pad's text or file
 * list changes, every client connected to that pad gets notified and
 * re-fetches the current state (avoids ordering issues — the server is
 * always the source of truth).
 *
 * The frontend automatically falls back to short-interval polling if the
 * WebSocket can't connect (e.g. a proxy that blocks upgrades).
 */

const { WebSocketServer } = require('ws');
const cookie = require('cookie');
const cookieSignature = require('cookie-signature');
const config = require('./config');
const { normalizePadId, getPad } = require('./services/padStore');
const { isPadUnlocked, isSiteAuthed: _unused } = require('./auth');

const rooms = new Map(); // padId -> Set<WebSocket>

function unsignCookie(raw) {
  if (!raw || !raw.startsWith('s:')) return null;
  const val = cookieSignature.unsign(raw.slice(2), config.cookieSecret);
  return val === false ? null : val;
}

function parseSignedCookies(cookieHeader) {
  const parsed = cookie.parse(cookieHeader || '');
  const out = {};
  for (const [k, v] of Object.entries(parsed)) {
    const decoded = decodeURIComponent(v);
    const unsigned = unsignCookie(decoded);
    if (unsigned !== null) out[k] = unsigned;
  }
  return out;
}

function isRequestAuthorized(req, padId) {
  const signed = parseSignedCookies(req.headers.cookie);
  if (config.sitePassword && signed.fp_site !== 'ok') return false;
  const pad = getPad(padId);
  if (pad && pad.password_hash) {
    let unlockedSet = new Set();
    try {
      const arr = JSON.parse(signed.fp_unlocked || '[]');
      if (Array.isArray(arr)) unlockedSet = new Set(arr);
    } catch (_) { /* cookie ausente/corrompido */ }
    const crypto = require('crypto');
    const hash = crypto.createHash('sha256').update(padId).digest('hex').slice(0, 16);
    if (!unlockedSet.has(hash)) return false;
  }
  return true;
}

function attach(server) {
  const wss = new WebSocketServer({ noServer: true });

  server.on('upgrade', (req, socket, head) => {
    let url;
    try {
      url = new URL(req.url, 'http://internal');
    } catch (_) {
      socket.destroy();
      return;
    }
    if (url.pathname !== '/ws') {
      socket.destroy();
      return;
    }
    const padId = normalizePadId(url.searchParams.get('pad') || '');
    if (!padId) {
      socket.destroy();
      return;
    }
    if (!isRequestAuthorized(req, padId)) {
      socket.write('HTTP/1.1 401 Unauthorized\r\n\r\n');
      socket.destroy();
      return;
    }
    wss.handleUpgrade(req, socket, head, (ws) => {
      ws.padId = padId;
      ws.isAlive = true;
      if (!rooms.has(padId)) rooms.set(padId, new Set());
      rooms.get(padId).add(ws);

      ws.on('pong', () => { ws.isAlive = true; });
      ws.on('close', () => {
        const room = rooms.get(padId);
        if (room) {
          room.delete(ws);
          if (room.size === 0) rooms.delete(padId);
        }
      });
      ws.on('error', () => ws.terminate());
    });
  });

  // Heartbeat: closes dead connections (e.g. a client that slept/lost network).
  const interval = setInterval(() => {
    for (const room of rooms.values()) {
      for (const ws of room) {
        if (ws.isAlive === false) {
          ws.terminate();
          continue;
        }
        ws.isAlive = false;
        ws.ping();
      }
    }
  }, 30000);
  server.on('close', () => clearInterval(interval));

  return wss;
}

/** Notifica todos os clientes ligados a um pad de que o estado mudou. */
function broadcastPadChanged(padId, extra = {}) {
  const room = rooms.get(padId);
  if (!room || room.size === 0) return;
  const payload = JSON.stringify({ type: 'changed', ...extra });
  for (const ws of room) {
    if (ws.readyState === ws.OPEN) {
      ws.send(payload);
    }
  }
}

module.exports = { attach, broadcastPadChanged };
